firedrake/projection.py

```diff
@@ -35,72 +81,129 @@ def project(v, V, bcs=None, mesh=None,
+        return element.entity_closure_dofs() == element.entity_dofs()
+
+    @cached_property
+    def A(self):
+        u = firedrake.TrialFunction(self.target.function_space())
+        v = firedrake.TestFunction(self.target.function_space())
+        a = firedrake.inner(u, v)*firedrake.dx
+        if self.is_DG and self.bcs is None:
+            a = firedrake.Tensor(a).inv
+        A = firedrake.assemble(a, bcs=self.bcs,
+                               form_compiler_parameters=self.form_compiler_parameters)
+        return A
+
+    @cached_property
+    def solver(self):
+        if self.is_DG and self.bcs is None:
+            class Dummy(object):
+                def __init__(self, A):
+                    self.A = A
+
+                def solve(self, x, b):
+                    self.A.force_evaluation()
+                    with x.dat.vec_wo as x_, b.dat.vec_ro as b_:
+                        self.A.M.handle.mult(b_, x_)
+                    return x
+
+            return Dummy(self.A)
+        else:
+            solver = firedrake.LinearSolver(self.A, solver_parameters=self.solver_parameters)
+            return solver
+
+    @property
+    def apply_massinv(self):
+        if not self.constant_jacobian:
+            firedrake.assemble(self.A.a, tensor=self.A, bcs=self.bcs,
+                               form_compiler_parameters=self.form_compiler_parameters)
+        return self.solver.solve
+
+    @cached_property
+    def residual(self):
+        return firedrake.Function(self.target.function_space())
+
+    @abc.abstractproperty
+    def rhs(self):
+        pass
+
+    def project(self):
+        self.apply_massinv(self.target, self.rhs)
+        return self.target
+
+
+class BasicProjector(ProjectorBase):
 
-    _solve(a == L, ret, bcs=bcs,
-           solver_parameters=solver_parameters,
-           form_compiler_parameters=form_compiler_parameters)
-    return ret
+    @cached_property
+    def rhs_form(self):
+        v = firedrake.TestFunction(self.target.function_space())
+        form = firedrake.inner(self.source, v)*firedrake.dx
+        return form
 
+    @cached_property
+    def assembler(self):
+        from firedrake.assemble import create_assembly_callable
+        return create_assembly_callable(self.rhs_form, tensor=self.residual,
+                                        form_compiler_parameters=self.form_compiler_parameters)
 
-class Projector(object):
+    @property
+    def rhs(self):
+        self.assembler()
+        return self.residual
+
+
+class SupermeshProjector(ProjectorBase):
+    @cached_property
+    def mixed_mass(self):
+        from firedrake.supermeshing import assemble_mixed_mass_matrix
+        return assemble_mixed_mass_matrix(self.source.function_space(),
+                                          self.target.function_space())
+
+    @property
+    def rhs(self):
+        with self.source.dat.vec_ro as u, self.residual.dat.vec_wo as v:
+            self.mixed_mass.mult(u, v)
```

Comment:
RHS is built by hitting the source vector with the mass matrix from the old space to the target space.  This is just integrating against a test function, but not supermeshing every time.